Update Getting started section from API docu (#2490)
wazuh/wazuh-documentation: source/user-manual/api/getting-started.rst

    # service wazuh-api start/status/stop/restart

User interface
--------------

.. versionadded:: 4.0.0

The Wazuh API interface is based on a free software project known as Swagger UI, which is very useful for listing and describing all available endpoints. It also allows you to interact with each of them, view example outputs and query all available parameters. By default, it can be found at ``https://localhost:55000/ui``.

.. note::
    If you don't provide your own, Wazuh API will create self-signed certificates. As a consequence, a security warning will appear in the browser. It is necessary to accept the risk in order to access the UI. For more information, please visit the :doc:`Securing the Wazuh API <../../installation-guide/securing_api>` section.

> The interface has a list of all available endpoints separated into groups (active-response, agents, ciscat, cluster, etc.). Each endpoint can have multiple methods (GET, POST, PUT, DELETE). To differentiate them, a specific color is used for each method, along with a label on the left. To run any endpoint, just click on it, press the *Try it out* button, fill in the necessary parameters and click on the *Execute* button. Please note that **login is required** to run endpoints. You can see how to do it in the next section, :ref:`Logging into the API <api_log_in>`.

Reviewer: ```suggestion
The interface has a list of all available endpoints separated into groups (active-response, agents, ciscat, cluster, etc.). Each endpoint can have multiple request methods (GET, POST, PUT, DELETE). To differentiate them, a specific color is used for each method, along with a label on the left. To run any endpoint, just click on it, press the *Try it out* button, fill in the necessary parameters and click on the *Execute* button. Please note that **login is required** to run endpoints. You can see how to do it in the next section, :ref:`Logging into the API <api_log_in>`.
```

Reviewer: add `request` word when talking about the GET, POST, PUT and DELETE methods

.. thumbnail:: ../../images/api/api-ui-overview.png
    :align: center
    :width: 100%

The endpoints can be used both in this interface and with the command-line tool *curl*, so in some examples we will use the two options.

.. _api_log_in:

Logging into the API
--------------------

Wazuh API endpoints require authentication in order to be used. Therefore, all calls must include a JSON Web Token. JWT is an open standard (RFC 7519) that defines a compact and self-contained way for securely transmitting information between parties as a JSON object.

> Logging into the UI

Reviewer: ```suggestion
Logging in with Swagger UI  
```

^^^^^^^^^^^^^^^^^^^^^^

1. Access the API interface at <protocol>://<host>:<port>/ui. The default path is ``https://localhost:55000/ui``.
2. Click on the green *Authorize* button.
3. Enter the username and password inside the section *basicAuth  (http, Basic)*. By default, the user is ``wazuh`` and the password is ``wazuh``.
4. Click on the green *Authorize* button under the password field and close this window.
5. Expand the endpoint ``GET /security/user/authenticate`` and click on the *Try it out* button.
6. Click on the *Execute* button. A result like the following will appear. Copy the value that appeared in ``<YOUR_JWT_TOKEN>``, without the quotes.

    .. code-block:: json
        :class: output

        {
          "token": "<YOUR_JWT_TOKEN>"
        }

7. Click again on the Authorize button.
8. Paste the value of the copied token into the *jwt  (http, Bearer)* field.

> 9. Click the Authorize button inside. After this, you can use any API endpoint allowed for the role to which the user belongs.

Reviewer: ```suggestion
9. Click on the Authorize button inside. After that, you can use any API endpoint allowed for the role the user belongs to.
```

Logging in with cURL
^^^^^^^^^^^^^^^^^^^^

> 1. Use the cURL command to get the JWT token. Replace <user> and <password> with yours. By default, the user is ``wazuh`` and the password is ``wazuh``.  If ``SSL`` (https) is enabled in the API and it is using the default self-signed certificates, it will be necessary to add the parameter ``-k`` as in the example above. Otherwise, it is not necessary to include ``-k``.

Reviewer: There is no example with `-k` above or below this sentence

    .. code-block:: console

        # curl -u <user>:<password> -X GET "https://localhost:55000/v4/security/user/authenticate"

    You will get a result similar to the following. Copy the token that you will find in ``<YOUR_JWT_TOKEN>`` without the quotes.

    .. code-block:: none
        :class: output

        {"token": "<YOUR_JWT_TOKEN>"}

> 2. Run any endpoint following this structure. Replace <endpoint> and <YOUR_JWT_TOKEN> with your values.

Reviewer: this shouldn't be step 2 but an actual example of how to use the token once the user has it.

    .. code-block:: console

        # curl -X <METHOD> "https://localhost:55000/v4/<ENDPOINT>" -H  "accept: application/json" -H  "Authorization: Bearer <YOUR_JWT_TOKEN>"

> Send a *request* to confirm that everything is working as expected:

Reviewer: this paragraph should be step 2 instead of the other one because shows how to check if previous step was successful

        "api_version": "4.0.0",
        "revision": 4000,
        "license_name": "GPL 2.0",
        "license_url": "https://github.com/wazuh/wazuh/blob/master/LICENSE",
        "hostname": "wazuh-master",
        "timestamp": "2020-05-25T07:05:00+0000"
    }

Basic concepts
--------------

Here are some of the basic concepts related to making API requests and understanding their responses:

-  The *cURL command* for each request contains:

    +-------------------------------------------------+----------------------------------------------------------------------------------------+
    | Field                                           | Description                                                                            |
    +=================================================+========================================================================================+
    | ``-X GET/POST/PUT/DELETE``                      | Specifies a custom request method to use when communicating with the HTTP server.      |
    +-------------------------------------------------+----------------------------------------------------------------------------------------+

>     | ``https://localhost:55000/v4/<ENDPOINT>``       | The API URL to use if you are running the command on the manager itself.               |

Reviewer: Expand this to specify the user must use HTTP or HTTPS depending on whether or not SSL is enabled and set up in the API.

- All responses are in *JSON format* and most of them follow this structure:

    +---------+----------------------+-----------------------------------------------------------------------+
    | Field   | Sub-field            | Description                                                           |
    +=========+======================+=======================================================================+
    | data    | affecte_items        | List with each one of the successfully affected items in the request. |
    |         +----------------------+-----------------------------------------------------------------------+
    |         | total_affected_items | Total number of successfully affected items.                          |
    |         +----------------------+-----------------------------------------------------------------------+
    |         | total_failed_items   | Total number of failed items.                                         |
    |         +----------------------+-----------------------------------------------------------------------+

>     |         | failed_items         | List with each one of the failed items in the request.                |

Reviewer: replace this
```
List with each one of the failed items in the request.
```
with:
```
List containing each of the failed items in the request.
```

.. note::
    If you don't provide your own, Wazuh API will create self-signed certificates. As a consequence, a security warning will appear in the browser. It is necessary to accept the risk in order to access the UI. For more information, please visit the :doc:`Securing the Wazuh API <../../installation-guide/securing_api>` section.

The interface has a list of all available endpoints separated into groups (active-response, agents, ciscat, cluster, etc.). Each endpoint can have multiple request methods (GET, POST, PUT, DELETE). To differentiate them, a specific color is used for each one, along with a label on the left. To run any endpoint, just click on it, press the *Try it out* button, fill in the necessary parameters and click on the *Execute* button. Please note that **login is required** to run endpoints. You can see how to do it in the next section, :ref:`Logging into the API <api_log_in>`.

.. thumbnail:: ../../images/api/api-ui-overview.png
    :align: center
    :width: 100%

> The endpoints can be used both in this interface and with the command-line tool *curl*, so in some examples we will use the two options.

Reviewer 2: ```suggestion
The Wazuh API endpoints can be reached via the swagger interface or using the command-line tool *curl*. In some of the examples below we will use both options.
```

Wazuh API endpoints require authentication in order to be used. Therefore, all calls must include a JSON Web Token. JWT is an open standard (RFC 7519) that defines a compact and self-contained way for securely transmitting information between parties as a JSON object.

Logging into the API UI
^^^^^^^^^^^^^^^^^^^^^^^

1. Access the API interface at <protocol>://<host>:<port>/ui. The default path is ``https://localhost:55000/ui``.
2. Click on the green *Authorize* button.
3. Enter the username and password inside the section *basicAuth  (http, Basic)*. By default, the user is ``wazuh`` and the password is ``wazuh``.
4. Click on the green *Authorize* button under the password field and close this window.
5. Expand the endpoint ``GET /security/user/authenticate`` and click on the *Try it out* button.
6. Click on the *Execute* button. A result like the following will appear. Copy the value that appeared in ``<YOUR_JWT_TOKEN>``, without the quotes.

    .. code-block:: json
        :class: output

        {
          "token": "<YOUR_JWT_TOKEN>"
        }

7. Click again on the Authorize button.
8. Paste the value of the copied token into the *jwt  (http, Bearer)* field.

> 9. Click on the Authorize button inside. After that, you can use any API endpoint allowed for the role the user belongs to.

Reviewer 2: ```suggestion
9. Click on the Authorize button inside. After that, you can use any API endpoint allowed for the roles the user belongs to.
```

            "failed_items": [
              {
                "error": {
                  "code": 1707,
                  "message": "Impossible to restart non-active agent: never_connected",
                  "remediation": "Please, make sure agent is active before attempting to restart"
                },
                "id": [
                  "001",
                  "002",
                ]
              },
              {
                "error": {
                  "code": 1707,
                  "message": "Impossible to restart non-active agent: disconnected",
                  "remediation": "Please, make sure agent is active before attempting to restart"
                },
                "id": [
                  "009",
                  "010"
                ]
              }
            ]
          },
          "message": "Could not send command to any agent"
        }

    - Example response to report a result error (code 400):

    .. code-block:: json
        :class: output

        {
          "type": "about:blank",
          "title": "Wazuh Error",
          "detail": "Permission denied: Resource type: *:*",
          "status": 400,

>           "remediation": "Please, make sure you have permissions to execute the current request. For more information on how to set up permissions, please visit XXXX",

Reviewer 2: Sustitute XXXX with the RBAC documentation link

          "code": 4000,
          "dapi_errors": {
            "master-node": {
              "error": "Permission denied: Resource type: *:*"
            }
          }
        }

    - Example response to report an unauthorized request (code 401):

    .. code-block:: json
        :class: output

        {
          "type": "about:blank",
          "title": "Unauthorized",
          "detail": "No authorization token provided",
          "status": 401
        }

- Responses containing collections of data will return a maximum of 500 elements. The *offset* and *limit* parameters may be used to iterate through large collections.
- All responses have an HTTP status code: 2xx (success), 4xx (client error), 5xx (server error), etc.
- All requests (except ``GET /`` and ``GET /security/user/authenticate``) accept the parameter ``pretty`` to convert the JSON response to a more human-readable format.

> - The API log is stored on the manager as ``/var/ossec/logs/api.log`` (the path and verbosity level can be changed in the API configuration file). The API logs are rotated daily. The rotations are stored in ``/var/ossec/logs/api/<year>/<month>`` and compressed using ``gzip``.

Reviewer 2: ```suggestion
- The API log is stored on the manager as ``/var/ossec/logs/api.log`` (the path and verbosity level can be changed in the API configuration file). The API logs are rotated daily. Rotated logs are stored in ``/var/ossec/logs/api/<year>/<month>`` and compressed using ``gzip``.
```

    .. code-block:: console

        # curl -u <user>:<password> -X GET "https://localhost:55000/v4/security/user/authenticate"

    You will get a result similar to the following. Copy the token that you will find in ``<YOUR_JWT_TOKEN>`` without the quotes.

    .. code-block:: none
        :class: output

        {"token": "<YOUR_JWT_TOKEN>"}

2. Send a *request* to confirm that everything is working as expected:

    .. code-block:: console

        curl -X GET "https://localhost:55000/v4/" -H "Authorization: Bearer <YOUR_JWT_TOKEN>"

    .. code-block:: json
        :class: output

        {
            "title": "Wazuh API",
            "api_version": "4.0.0",
            "revision": 4000,
            "license_name": "GPL 2.0",
            "license_url": "https://github.com/wazuh/wazuh/blob/master/LICENSE",
            "hostname": "wazuh-master",
            "timestamp": "2020-05-25T07:05:00+0000"
        }

> Run any endpoint following the structure below. Replace <endpoint> and <YOUR_JWT_TOKEN> with your values.

Reviewer: Update this line to add something like `Once we are logged in we can run any endpoint following the structure below.`

    +-------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    | ``-k``                                          | Suppress SSL certificate errors (only if you use the default self-signed certificates).                                                                            |
    +-------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+

- All responses are in *JSON format* and most of them follow this structure:

    +---------+----------------------+-----------------------------------------------------------------------+
    | Field   | Sub-field            | Description                                                           |
    +=========+======================+=======================================================================+

>     | data    | affecte_items        | List with each one of the successfully affected items in the request. |

Reviewer: There is a typo in `affected_items` sub-field
